### PR TITLE
avm2: Add automatic linting for AS3 playerglobals

### PR DESCRIPTION
**Pull request by EmperorBale**
This PR adds linting to AS3 playerglobals for better code quality. It is using FlexPMD behind the scenes, which requires java, which is not a big deal since building playerglobals also requires java.

I added a new crate called `as3lint`, which can be used to lint all the code in playerglobals. To use it, enter the following command: `cargo run -p as3lint`. 

I also added running `as3lint` as part of the automatic workflow, which will work the same way as checking formatting & clippy.

The linter can also be customized to our liking. Go to `as3lint/flexpmd/ruffle-ruleset.xml` to find the XML file containing all the active rules. 
Heres an example of what a rule would look like:
```
<rule since="" class="com.adobe.ac.pmd.rules.architecture.MonkeyPatchingRule" message="This class looks to be duplicated with a SDK class">
    <description>Monkey patching can be a risky undertaking because it is not using intended extensibility points and thus may have unintended consequences or make migration to newer versions of the SDK more difficult</description>
    <priority>1</priority>
  </rule>
```
From the XML file, we can customize or remove the rule. This is important, because a lot of the rules that come by default are invalid for playerglobals, or make it impossible to implement certain classes. I have already removed majority of problematic rules, but if you find any that I missed, feel free to remove them.

### CODE REVIEW

**Comment by EmperorBale** [comment]
For an example of what a linting failure will look like, see https://github.com/ruffle-rs/ruffle/runs/7825376567?check_suite_focus=true#step:6:98

**Review by Aaron1011 — APPROVED** [review]

**Comment by relrelb** [comment]
IMO this can be hoisted into the `build_playerglobal` crate, without introducing a dedicated crate for the linter.

File: .github/workflows/test_rust.yml [hunk 1]
```diff
@@ -64,6 +64,12 @@ jobs:
           command: fmt
           args: --all -- --check
 
+      - name: Lint AS3 playerglobals 
+        uses: actions-rs/cargo@v1
```
Comment (adrian17) [inline .github/workflows/test_rust.yml:68]: Let's not run this on entire matrix, linux-only or even linux-stable-only should be enough.

**Comment by EmperorBale** [comment]
> IMO this can be hoisted into the `build_playerglobal` crate, without introducing a dedicated crate for the linter.

Could you provide more info as to how this would work? Personally, I like it being a separate crate because I can run it using just `cargo run -p as3lint`.

**Comment by danielhjacobs** [comment]
May be a good idea to add this command to https://github.com/ruffle-rs/ruffle/blob/master/CONTRIBUTING.md#code-guidelines or maybe just https://github.com/ruffle-rs/ruffle/blob/master/core/src/avm2/globals/README.md

**Comment by relrelb** [comment]
> > IMO this can be hoisted into the `build_playerglobal` crate, without introducing a dedicated crate for the linter.
> 
> Could you provide more info as to how this would work? Personally, I like it being a separate crate because I can run it using just `cargo run -p as3lint`.

I thought the existing `build_playerglobal` facility can be reused for compiling tests before running them, so we can avoid `.fla` files where possible. For the same reason, the AS3 linter can be used to lint tests. I guess it could best fit as a toplevel  `compiler` crate that will support compilation and lint of both AVM1 and AVM2 code, and eventually even SWF tags generation based on our own DSL. With that in mind, for now I suggest adding the AS3 linter to `core/build_playerglobal`, and bind it to the CLI so it still can be invoked as something like `cargo run -p core/build_playerglobal lint [files...]`.

**Comment by EmperorBale** [comment]
Okay, so I moved as3lint to build_playerglobal, but there are still a few things up for discussion. First of all, the new CLI syntax is:
`cargo run -p build_playerglobal -- lint` - for linting,
`cargo run -p build_playerglobal -- compile <OUT_DIR>` - for compiling

The compiling cli is a little different - we no longer have control over the root directory. Instead, i've decided to just have it auto detect it instead, which should be fine as long as `build_playerglobal` is only ever used inside of Ruffle. Linting is also using this auto detecting method. I think this is for the best, because the command for linting should ideally "just work" without needing to manually supply the path to globals, making it as easy as possible for people trying to add new AS3 code, but let me know if you have a better idea.

Another thing I wanted to discuss was possibly renaming the `build_playerglobal` crate to just `playerglobal`, since I think it makes more sense with the new CLI options.

**Comment by Herschel** [comment]
I'm a little worried about the cost-benefit on this one -- it's another not-really-maintained Java tool, and I have a feeling we're going to end up fighting this more than it helps. That said, seeing that the lints are configurable, I'm willing to give it a shot! I do like that it's already caught some unnecessary imports.

**Comment by Aaron1011** [comment]
Is there an easy way to suppress lints per file/directory if necessaey

**Comment by EmperorBale** [comment]
> Is there an easy way to suppress lints per file/directory if necessary?

Not currently

**Comment by Herschel** [comment]
Okay, let's merge this so we're not constantly rebasing it as new AS3 code gets added. :-) Removed `EventMissingCloneFunctionRule` lint because quite a few of the AS3 events don't override `clone`.

I think what I'd really like eventually is a nice AS3 formatter.

Thank you!

**Pull request merged**